src/abi/solana_accounts.rs

```diff
@@ -0,0 +1,414 @@
+    }
+}
+
+/// Collect the accounts this contract needs
+pub(super) fn collect_accounts_from_contract(
+    contract_no: usize,
+    ns: &Namespace,
+) -> HashMap<usize, IndexSet<SolanaAccount>> {
+    let mut visiting_queue: IndexSet<usize> = IndexSet::from_iter(
+        ns.contracts[contract_no]
+            .functions
+            .iter()
+            .map(|ast_no| ns.contracts[contract_no].all_functions[ast_no]),
+    );
+
+    let contract = &ns.contracts[contract_no];
+    let mut recurse_data = RecurseData {
+        accounts: HashMap::new(),
+        next_queue: IndexSet::new(),
+        cfg_func_no: 0,
+        accounts_added: 0,
+        ns,
+    };
+
+    let mut old_size: usize = 0;
+    loop {
+        for item in &visiting_queue {
+            if *item == usize::MAX {
+                continue;
+            }
+
+            recurse_data.cfg_func_no = *item;
+            check_function(&contract.cfg[*item], &mut recurse_data);
+        }
+
+        // This is the convergence condition for this loop.
+        // If we have not added new accounts to the map in this iteration and the queue for the
+        // next iteration does not have any new element, we can stop.
+        if old_size == recurse_data.accounts_added
+            && visiting_queue.len() == recurse_data.next_queue.len()
+        {
+            break;
+        }
+        old_size = recurse_data.accounts_added;
+        std::mem::swap(&mut visiting_queue, &mut recurse_data.next_queue);
+        recurse_data.next_queue.clear();
+    }
+
+    recurse_data.accounts
+}
+
+/// Collect the accounts in a function
+fn check_function(cfg: &ControlFlowGraph, data: &mut RecurseData) {
+    if cfg.blocks.is_empty() {
+        return;
+    }
+    let mut queue: VecDeque<usize> = VecDeque::new();
+    let mut visited: HashSet<usize> = HashSet::new();
+    queue.push_back(0);
+    visited.insert(0);
+
+    while let Some(cur_block) = queue.pop_front() {
+        for instr in &cfg.blocks[cur_block].instr {
+            check_instruction(instr, data);
+        }
+        // TODO: Block edges is an expensive function, we use it six time throughout the code,
+        // perhaps we can just use the dag I calculate during cse.
+        // Changes in constant folding would be necessary
+        for edge in cfg.blocks[cur_block].edges() {
+            if !visited.contains(&edge) {
+                queue.push_back(edge);
+                visited.insert(edge);
+            }
+        }
+    }
+}
+
+/// Collect the accounts in an instruction
+fn check_instruction(instr: &Instr, data: &mut RecurseData) {
+    match instr {
+        Instr::Print { expr }
+        | Instr::LoadStorage { storage: expr, .. }
+        | Instr::ClearStorage { storage: expr, .. }
+        | Instr::BranchCond { cond: expr, .. }
+        | Instr::PopStorage { storage: expr, .. }
+        | Instr::SelfDestruct { recipient: expr }
+        | Instr::Set { expr, .. } => {
+            expr.recurse(data, check_expression);
```

Comment:
No, because I gather information from the Expressions themselves. It would be possible if I create a data structure to keep track of the instruction the expression belongs to and the recursion depth for the Expression, but this is too much of a trick.